UIforETW/Utility.cpp

```diff
@@ -20,7 +20,133 @@ limitations under the License.
+	}
+}
+
+std::wstring GetDocumentsFolderPath()
+{
+	ATL::CComHeapPtr<_Null_terminated_ wchar_t> docsPathTemp;
+	const HRESULT docsPathResult =
+		::SHGetKnownFolderPath(
+		FOLDERID_Documents, KF_FLAG_NO_ALIAS, NULL, &docsPathTemp.m_pData);
+	if (FAILED(docsPathResult))
+	{
+		debugPrintf(L"SHGetKnownFolderPath (for Documents) failed to retrieve the path.\n");
+		std::terminate();
+	}
+	return docsPathTemp.m_pData;
+}
+
+void copyWPAProfileToDocuments(const bool force)
+{
+	// First copy the WPA 8.1 startup.wpaProfile file
+	std::wstring docsPath(GetDocumentsFolderPath());
+
+	if (force)
+		outputPrintf(L"\n");
+
+	if (docsPath.empty())
+	{
+		outputPrintf( L"Failed to copy WPA profile to documents. See debugger output for details.\n" );
+		return;
+	}
+
+	const std::wstring source = docsPath + kWPAStartupFileName;
+	const std::wstring destDir = docsPath + std::wstring(L"\\WPA Files");
+	const std::wstring dest = destDir + kWPAStartupFileName;
+	const BOOL destinationExists = ::PathFileExistsW(dest.c_str());
+	if (force || !destinationExists)
+	{
+		ATLVERIFY(::CreateDirectoryW(destDir.c_str(), NULL));
+		const BOOL copyResult = ::CopyFileW(source.c_str(), dest.c_str(), FALSE);
+		if (copyResult)
+		{
+			if (force)
+				outputPrintf(L"Copied Startup.wpaProfile to the WPA Files directory.\n");
+			return;
+		}
+		if (force)
+		{
+			outputPrintf(L"Failed to copy Startup.wpaProfile to the WPA Files directory.\n");
+			outputLastError();
+			return;
+		}
+	}
+}
+
+void copyWPAProfileToLocalAppData(const std::wstring& exeDir, const bool force)
+{
+	PCWSTR const localAppDataEnvVar = L"localappdata";
+	// Then copy the WPA 10 startup.wpaProfile file
+	const std::wstring localAppData = GetEnvironmentVariableString(localAppDataEnvVar);
+	if (localAppData.empty())
+	{
+		outputPrintf(L"the `%s` environment variable didn't contain a valid path. Failed to copy WPA 10 profile.\n", localAppDataEnvVar);
+		return;
+	}
+	std::wstring source = exeDir + L"\\startup10.wpaProfile";
+	std::wstring destDir = std::wstring(localAppData) + L"\\Windows Performance Analyzer";
+	std::wstring dest = destDir + kWPAStartupFileName;
+	if (force || !::PathFileExistsW(dest.c_str()))
+	{
+
+		ATLVERIFY(::CreateDirectoryW(destDir.c_str(), NULL));
+
+		if (::CopyFileW(source.c_str(), dest.c_str(), FALSE))
+		{
+			if (force)
+				outputPrintf(L"%s", L"Copied Startup.10wpaProfile to %localappdata%\\Windows Performance Analyzer\n");
+			return;
+		}
+		if (force)
+		{
+			outputPrintf(L"%s", L"Failed to copy Startup.10wpaProfile to %localappdata%\\Windows Performance Analyzer\n");
+			outputLastError();
+		}
+	}
+
+}
+} //namespace {
+
+void outputLastError(const DWORD lastErr)
+{
+	const DWORD errMsgSize = 1024u;
+	wchar_t errBuff[errMsgSize] = {0};
+	const DWORD ret = ::FormatMessageW(
+		(FORMAT_MESSAGE_FROM_SYSTEM | FORMAT_MESSAGE_IGNORE_INSERTS),
+		NULL, lastErr, MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT),
+		errBuff, errMsgSize, NULL);
+
+	if (ret == 0)

```

Comment:
Why should the program terminate if there is a problem with printing an error message?